File: iosApp/iosApp/Components/AvatarStack.swift
import SwiftUI

/// Minimal member shape `AvatarStack` needs — decoupled from any specific UI model.
struct AvatarStackMember: Identifiable {
    let id: String
    let name: String
    let avatarUrl: String?
}

/// Overlapping row of avatars with a "+N" overflow chip. Mirrors web's `AvatarStack` /
/// Android's `AvatarStack`: first avatar on top, ~1/3-size overlap, up to `max` shown.
struct AvatarStack: View {
    let members: [AvatarStackMember]
    var size: CGFloat = 24
    var max: Int = 3
    var ringColor: Color = Color(uiColor: .systemBackground)

    private var shown: [AvatarStackMember] { Array(members.prefix(max)) }
    private var extra: Int { members.count - shown.count }
    private var overlap: CGFloat { size / 3 }

    var body: some View {
        HStack(spacing: 0) {
            ForEach(Array(shown.enumerated()), id: \.element.id) { index, member in
                Avatar(name: member.name, avatarUrl: member.avatarUrl, size: size, memberId: member.id)
                    .overlay(Circle().strokeBorder(ringColor, lineWidth: 2))
                    .zIndex(Double(shown.count - index))
                    .offset(x: -overlap * CGFloat(index))
            }
            if extra > 0 {
                ZStack {
                    Circle().fill(Color.avatarStackOverflowBg)
                    Text("+\(extra)")
                        .font(.plexSansMedium(size: size * 0.4))
                        .foregroundColor(.white)
                }
                .frame(width: size, height: size)
                .overlay(Circle().strokeBorder(ringColor, lineWidth: 2))
                .zIndex(0)
                .offset(x: -overlap * CGFloat(shown.count))
            }
        }
        .padding(.trailing, extra > 0 ? overlap * CGFloat(shown.count) : overlap * CGFloat(Swift.max(shown.count - 1, 0)))
    }
}

#Preview {
    AvatarStack(members: [
        AvatarStackMember(id: "1", name: "Ana Silva", avatarUrl: nil),
        AvatarStackMember(id: "2", name: "Ben Choi", avatarUrl: nil),
        AvatarStackMember(id: "3", name: "Cara Doyle", avatarUrl: nil),
        AvatarStackMember(id: "4", name: "Dev Patel", avatarUrl: nil),
    ])
    .padding()
}

File: designsystem/src/androidMain/kotlin/com/ivangarzab/kluvs/designsystem/components/avatars/AvatarStack.kt
package com.ivangarzab.kluvs.designsystem.components.avatars

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.PreviewLightDark
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.zIndex
import com.ivangarzab.kluvs.designsystem.theme.KluvsTheme
import com.ivangarzab.kluvs.designsystem.theme.avatarStackOverflowBg

/** Minimal member shape [AvatarStack] needs — decoupled from any specific UI model. */
data class AvatarStackMember(
    val id: String,
    val name: String,
    val avatarUrl: String?,
)

/**
 * Overlapping row of avatars with a "+N" overflow chip. Mirrors web's `AvatarStack`:
 * first avatar on top, -8px overlap (at 24dp size), up to 3 shown, ring matching the
 * surrounding surface.
 */
@Composable
fun AvatarStack(
    members: List<AvatarStackMember>,
    modifier: Modifier = Modifier,
    size: Dp = 24.dp,
    max: Int = 3,
    ringColor: Color = KluvsTheme.colors.background,
) {
    val shown = members.take(max)
    val extra = members.size - shown.size
    val overlap = size / 3f // ~ -8px at 24dp per design-system spec

    Row(modifier = modifier) {
        shown.forEachIndexed { index, member ->
            Avatar(
                name = member.name,
                memberId = member.id,
                avatarUrl = member.avatarUrl,
                size = size,
                modifier = Modifier
                    .offset(x = -overlap * index)
                    .zIndex((shown.size - index).toFloat())
                    .border(width = 2.dp, color = ringColor, shape = CircleShape)
            )
        }
        if (extra > 0) {
            Box(
                modifier = Modifier
                    .offset(x = -overlap * shown.size)
                    .zIndex(0f)
                    .size(size)
                    .clip(CircleShape)
                    .background(color = avatarStackOverflowBg)
                    .border(width = 2.dp, color = ringColor, shape = CircleShape),
                contentAlignment = Alignment.Center
            ) {
                Text(
                    text = "+$extra",
                    color = Color.White,
                    fontWeight = FontWeight.Medium,
                    fontSize = (size.value * 0.4f).sp
                )
            }
        }
    }
}

@PreviewLightDark
@Composable
fun Preview_AvatarStack() = KluvsTheme {
    Box(
        modifier = Modifier
            .background(color = KluvsTheme.colors.background)
    ) {
        AvatarStack(
            members = listOf(
                AvatarStackMember("1", "Ana Silva", null),
                AvatarStackMember("2", "Ben Choi", null),
                AvatarStackMember("3", "Cara Doyle", null),
                AvatarStackMember("4", "Dev Patel", null),
                AvatarStackMember("5", "Eve Kim", null),
            )
        )
    }
}
